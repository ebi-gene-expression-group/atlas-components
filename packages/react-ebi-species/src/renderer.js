"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
