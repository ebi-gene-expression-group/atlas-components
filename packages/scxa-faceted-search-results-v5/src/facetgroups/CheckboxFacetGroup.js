import React from 'react'
import PropTypes from 'prop-types'

import TooltipIcon from './TooltipIcon'
import FacetGroupPropTypes from "./FacetGroupPropTypes";

const CheckboxOption = ({group, value, label, disabled, checked, onChange}) =>
  <div>
    <input type={`checkbox`} {...{value, checked, disabled}}
           onChange={() => onChange({label, group, disabled})}/>
    <label style={disabled ? {color: `lightgrey`} : {}}>{label}</label>
  </div>

CheckboxOption.propTypes = {
  group: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired
}

class CheckboxFacetGroup extends React.Component {

  getSelectedFacets(checkedFacets, changedFacetLabel) {
    if (checkedFacets.indexOf(`true`) !== -1) {
      checkedFacets = this.props.facets.map(facet => facet.label)
    }

    if (checkedFacets.length === 0) {
      checkedFacets.push(changedFacetLabel)
    } else {
      if (checkedFacets.indexOf(changedFacetLabel) !== -1) {
        checkedFacets.splice(checkedFacets.indexOf(changedFacetLabel), 1)
      } else {
        checkedFacets.push(changedFacetLabel)
      }
    }

    return this.props.facets.filter( facet => checkedFacets.includes(facet.label))
  }
  render() {
    const { name, description, facets, queryParams } = this.props
    const checkedFacets = queryParams

    return (facets.length > 0 && (<div id={`facetGroupCheckBox`} className={`padding-bottom-xlarge`}>
        <h4>
          {name}
          {description && <TooltipIcon tooltipText={description}/>}
        </h4>
        {facets.map((facet) =>
          <CheckboxOption {...facet}
                          checked={
                            checkedFacets.some((checkedFacet) => checkedFacet === `true` || checkedFacet === facet.label)}
                          onChange={(changedFacet) =>
                            this.props.onChange(changedFacet.group,
                              this.getSelectedFacets(checkedFacets, changedFacet.label), this.props.facets)}
                          key={facet.label}
          />
        )}
      </div>
    ))

  }
}

CheckboxFacetGroup.propTypes = FacetGroupPropTypes

export default CheckboxFacetGroup
