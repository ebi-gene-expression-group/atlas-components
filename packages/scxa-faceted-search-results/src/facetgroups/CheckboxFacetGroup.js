import React from 'react'
import PropTypes from 'prop-types'
import { xorBy as _xorBy } from 'lodash'

import FacetGroupPropTypes from './FacetGroupPropTypes'
import TooltipIcon from './TooltipIcon'

const CheckboxOption = ({ group, value, label, disabled, checked, onChange }) =>
  <div>
    <input type={`checkbox`} {...{value, checked, disabled}}
      onChange={() => onChange({group, label, value, disabled})}/>
    <label style={disabled ? {color: `lightgrey`} : {}}>{label}</label>
  </div>

CheckboxOption.propTypes = {
  group: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired
}

// In principle we donâ€™t need this component to be stateful, but in doing so we can create a custom _handleChange
// function that will ultimately call onChange(facetGroupName, facets); this allows us to have the same API as
// React-Select and reuse the same callback for both checkbox-style and multiselect facet groups
class CheckboxFacetGroup extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      checkedFacets: []
    }
    this._handleChange = this._handleChange.bind(this)
  }

  // disabled is passed in but not used (it makes tests easier)
  _handleChange(facet) {
    this.setState(
      { checkedFacets: _xorBy(this.state.checkedFacets, [facet], `value`) },
      () => this.props.onChange(facet.group, this.state.checkedFacets))
  }

  render() {
    const { facetGroupName, facetGroupNameDescription, facets } = this.props
    const { checkedFacets } = this.state

    return (
      <div className={`padding-bottom-xlarge`}>
        <h4>
          { facetGroupName }
          { facetGroupNameDescription && <TooltipIcon tooltipText={facetGroupNameDescription}/> }
        </h4>
        {facets.map((facet) =>
          <CheckboxOption {...facet}
            checked={checkedFacets.some((checkedFacet) => checkedFacet.value === facet.value)}
            onChange={this._handleChange}
            key={facet.value}/>
        )}
      </div>
    )
  }
}

CheckboxFacetGroup.propTypes = FacetGroupPropTypes

export default CheckboxFacetGroup
