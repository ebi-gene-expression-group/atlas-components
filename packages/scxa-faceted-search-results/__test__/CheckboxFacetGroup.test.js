import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import CheckboxFacetGroup from '../src/facetgroups/CheckboxFacetGroup'

Enzyme.configure({ adapter: new Adapter() })

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min; //The maximum is exclusive and the minimum is inclusive
}

describe(`CheckboxFacetGroup`, () => {
  const vindicators = [
    { value: `supernova`, label: `Supernova`},
    { value: `noob-noob`, label: `Noob-Noob` },
    { value: `vance_maximus`, label: `Vance Maximus`},
    { value: `alan_rails`, label: `Alan Rails`},
    { value: `crocubot`, label: `Crocubot` },
    { value: `million_ants`, label: `Million Ants`},
    { value: `morty_smith`, label: `Morty Smith`},
    { value: `rick_sanchez`, label: `Rick Sanchez`},
    { value: `lady_katana`, label: `Lady Katana`},
    { value: `calypso`, label: `Calypso`},
    { value: `diablo_verde`, label: `Diablo Verde`}
  ]

  const props = {
    facetName: `Vindicators`,
    hideName: false,
    facetItems: vindicators.filter((vindicator) => Math.random() > 0.5),
    onChange: () => {}
  }

  test(`renders the right number of checkboxes and the facet name`, () => {
    const wrapper = mount(<CheckboxFacetGroup {...props} />)
    expect(wrapper.find(`input`)).toHaveLength(props.facetItems.length)
    expect(wrapper.find({ type: `checkbox` })).toHaveLength(props.facetItems.length)
    expect(wrapper.find(`h4`)).toHaveLength(1)
    expect(wrapper.find(`h4`).text()).toEqual(props.facetName)
  })

  test(`can hide the header`, () => {
    const wrapper = mount(<CheckboxFacetGroup {...props} hideName={true} />)
    expect(wrapper.find(`h4`)).toHaveLength(0)
  })

  test(`callback is called when a checkbox is checked/unchecked with the right arguments`, () => {
    const randomCheckboxIndex = getRandomInt(0, props.facetItems.length)
    const mockCallback = jest.fn()
    const wrapper = mount(<CheckboxFacetGroup {...props} onChange={mockCallback} />)
    wrapper.find({ type: `checkbox` }).at(randomCheckboxIndex).simulate(`change`)
    expect(mockCallback.mock.calls.length).toBe(1)
    expect(mockCallback.mock.calls[0]).toEqual([`Vindicators`, [props.facetItems[randomCheckboxIndex]]])
    wrapper.find({ type: `checkbox` }).at(randomCheckboxIndex).simulate(`change`)
    expect(mockCallback.mock.calls.length).toBe(2)
    expect(mockCallback.mock.calls[1]).toEqual([`Vindicators`, []])
  })

  test(`matches snapshot`, () => {
    const tree = renderer.create(<CheckboxFacetGroup {...props} facetItems={vindicators}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
