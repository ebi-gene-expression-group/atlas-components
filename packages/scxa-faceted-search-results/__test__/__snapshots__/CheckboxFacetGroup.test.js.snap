// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxFacetGroup matches snapshot 1`] = `
<div
  className="padding-bottom-xlarge"
>
  <h4>
    Planet
    <sup>
       
      <span
        className="icon icon-generic"
        data-html={true}
        data-icon="i"
        data-tip="<span>Planets where events in this episode take<br>place</span>"
        style={
          Object {
            "color": "lightgrey",
            "fontSize": "smaller",
          }
        }
      />
    </sup>
  </h4>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="gazorpazorp"
    />
    <label
      style={Object {}}
    >
      Gazorpazorp
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="squanch"
    />
    <label
      style={Object {}}
    >
      Squanch
    </label>
  </div>
  <div>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
      value="buttworld"
    />
    <label
      style={Object {}}
    >
      Buttworld
    </label>
  </div>
</div>
`;
